lib/srv/heartbeatv2_test.go

```diff
@@ -459,3 +460,77 @@ func awaitEvents(t *testing.T, ch <-chan hbv2TestEvent, opts ...eventOption) {
 		}
 	}
 }
+
+type fakeDownstreamHandle struct {
+	inventory.DownstreamHandle
+}
+
+func (f *fakeDownstreamHandle) CloseContext() context.Context {
+	return context.Background()
+}
+
+func mockMetadataGetter() (metadataGetter, chan *metadata.Metadata) {
+	ch := make(chan *metadata.Metadata, 1)
+	return func(ctx context.Context) (*metadata.Metadata, error) {
+		meta := <-ch
+		if meta == nil {
+			return nil, fmt.Errorf("error fetching metadata")
+		}
+		return meta, nil
+	}, ch
+}
+
+func makeMetadata(id string) *metadata.Metadata {
+	return &metadata.Metadata{
+		CloudMetadata: &types.CloudMetadata{
+			AWS: &types.AWSInfo{
+				InstanceID: id,
+			},
+		},
+	}
+}
+
+func TestNewHeartbeatFetchMetadata(t *testing.T) {
+	t.Parallel()
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	heartbeat, err := NewSSHServerHeartbeat(SSHServerHeartbeatConfig{
+		InventoryHandle: &fakeDownstreamHandle{},
+		GetServer: func() *types.ServerV2 {
+			return &types.ServerV2{
+				Spec: types.ServerSpecV2{},
+			}
+		},
+	})
+	require.NoError(t, err)
+	metadataGetter, metaCh := mockMetadataGetter()
+	t.Cleanup(func() { close(metaCh) })
+	inner := heartbeat.inner.(*sshServerHeartbeatV2)
+	inner.getMetadata = metadataGetter
+
+	// Metadata won't be set before metadata getter returns.
+	server := inner.getServer(ctx)
+	require.Nil(t, server.GetCloudMetadata())
+
+	// Metadata won't be set if the getter fails.
+	metaCh <- nil
+	require.Never(t, func() bool {
+		meta := inner.getServer(ctx).GetCloudMetadata()
+		return meta != nil
+	}, 100*time.Millisecond, 10*time.Millisecond)
```

Comment:
Do the assert without the require.Never, so we don't always block to the end?

I realize there's a race, but this almost like sleeping. (TBH I'd rather have the sleep, as it's clearer.)

```suggestion
	time.Sleep(100*time.Millisecond) // Wait for goroutines to complete.
	assert.Nil(t, inner.getServer(ctx).GetCloudMetadata())
```

Another option is to give a short grace period for the goroutine, so it hits more consistently in the first try (and then we can test assuming it's synchronous).